src/lib/device/hs_super_blk.h

```diff
@@ -0,0 +1,189 @@
+
+public:
+    uint64_t gen_number{0};                   // Generation count of this structure
+    uint32_t version{0};                      // Version Id of this structure
+    char product_name[s_product_name_size]{}; // Product name
+
+    uint32_t num_pdevs{0};                    // Total number of pdevs homestore is being created on
+    uint32_t max_vdevs{0};                    // Max VDevs possible, this cannot be changed post formatting
+    uint32_t max_system_chunks{0};            // Max Chunks possible, this cannot be changed post formatting
+
+    uint64_t system_uuid{0};                  // homestore system uuid.
+
+public:
+    const char* get_product_name() const { return product_name; }
+    uint32_t get_version() const { return version; }
+    void set_system_uuid(const hs_uuid_t uuid) { system_uuid = static_cast< uint64_t >(uuid); }
+    hs_uuid_t get_system_uuid() const { return static_cast< hs_uuid_t >(system_uuid); }
+
+    std::string to_string() const {
+        auto str = fmt::format("gen_number={}, version={}, product_name={} system_uuid={}", gen_number, get_version(),
+                               get_product_name(), get_system_uuid());
+        return str;
+    }
+};
+
+struct pdev_info_header {
+public:
+    uint64_t data_offset{0};         // Offset within pdev where data starts
+    uint64_t size{0};                // Total pdev size
+    uint32_t pdev_id{0};             // Device ID for this store instance.
+    uint32_t max_pdev_chunks{0};     // Max chunks in this pdev possible
+    disk_attr dev_attr;              // Attributes homestore expects from all the devices.
+    uint8_t mirror_super_block{0x0}; // Have we mirrored the super block on head/tail
+
+public:
+    std::string to_string() const {
+        auto str =
+            fmt::format("data_offset={}, size={}, pdev_id={} max_pdev_chunks={} dev_attr=[{}] mirror_super_block?={}",
+                        in_bytes(data_offset), in_bytes(size), pdev_id, max_pdev_chunks, dev_attr.to_string(),
+                        (mirror_super_block == 0x00) ? "false" : "true");
+        return str;
+    }
+};
+
+struct first_block {
+    static constexpr uint32_t s_atomic_fb_size{512};       // increase 512 to actual size if in the future first_block
+                                                           // can be larger;
+    static constexpr uint32_t s_io_fb_size{4096};          // This is the size we do IO on, with padding
+    static constexpr uint32_t HOMESTORE_MAGIC{0xCEEDDEEB}; // Magic written as first bytes on each device
+
+public:
+    uint64_t magic{0};              // Header magic expected to be at the top of block
+    uint32_t checksum{0};           // Checksum of the entire first block (excluding this field)
+    first_block_header hdr;         // Information about the entire system
+    pdev_info_header this_pdev_hdr; // Information about the current pdev
+
+public:
+    uint64_t get_magic() const { return magic; }
+
+    bool is_valid() const {
```

Comment:
we should also check version, right? We are not interested with drives formatted with old version homestore stuff, in other words, it will be booted as first time boot if it is an old version.

Reply:
We are checking the version inside first_block_header. This first 2 check is to see if they are indeed any homestore device or not.